doc/man/gnunet-pseudonym.1
.TH GNUNET-PSEUDONYM "1" "6 Aug 2013" "GNUnet"
.SH NAME
gnunet\-pseudonym \- advertise or list pseudonyms
.SH SYNOPSIS
.B gnunet\-pseudonym
[options]
.SH DESCRIPTION
.PP
gnunet\-pseudonym is a tool for advertising pseudonyms.  A pseudonym is the persona that controls a namespace.  As such, it is identical to a public\-private ECC key pair.  A namespace is a collection of files that have been signed by the corresponding private ECC key.  A namespace is typically associated with a nickname and other metadata.

Namespaces are an important tool for providing assurances about content integrity and authenticity in GNUnet.  Since all of the content in the namespace must have been provided by the same entity, users can form an opinion about that entity and learn to search (or avoid) certain namespaces.

gnunet\-pseudonym can be used to list all of the pseudonyms that were created locally, to advertise namespaces to other users, and to list all of the namespaces (with their meta\-data) known to the local user.  By default, gnunet\-pseudonym lists all pseudonyms that were discovered so far.

Advertising a pseudonym requires using the \-A option together with the name of the pseudonym.  Names for pseudonyms controlled by the user must be unique.  However, names for namespaces of other users may not be globally unique.  If two foreign namespaces in GNUnet use the same nickname, all GNUnet tools will display the nickname together with a number which ensures that the name becomes locally unique to avoid ambiguity.  Additional options can be passed together with the \-A option to provide additional meta\-data that describes the namespace.  Possible meta\-data includes the 'realname' of the person controlling the namespace, a description, the mime\-type for content in the namespace (useful if the namespace is dedicated to some specific type of content) and contact information.  One important piece of meta\-data that can be specified is the identifier of a document root, that is the name of a file in the namespace that is a portal to the rest of the content.  This is useful to help users find this root in the absence of conventions.  Note that all of this meta\-data is optional and should never be trusted blindly.

As mentioned before, by default, gnunet\-pseudonym simply lists the meta\-data available for other namespaces.  Namespaces can be discovered whenever the peer obtains the namespace advertisement.  Namespace advertisements can be found using ordinary keyword\-based searches (by default gnunet\-pseudonym publishes the namespace advertisement under the keyword 'namespace', but the \-k option can be used to specify other keywords) and under the 'empty' identifier of the respective namespace (using a namespace\-search if the namespace ID is already known).

For more details about GNUnet namespaces and content encoding please read the 'Encoding for Censorship\-resistant Sharing' (ECRS) paper which can be found on the GNUnet webpage.

.TP
\fB\-a \fILEVEL\fR, \fB\-\-anonymity=LEVEL\fR
set desired level of sender anonymity.  Default is 1.

.TP
\fB\-A NAME\fR, \fB\-\-advertise=NAME\fR
Advertise namespace of the pseudonym with the given NAME.

.TP
\fB\-h\fR, \fB\-\-help\fR
Print help page.

.TP
\fB\-k KEYWORD\fR, \fB\-\-keyword=KEYWORD\fR
Publish a namespace advertisement under the keyword 'KEYWORD'.  Default is 'namespace' (use with \-C).  You can specify \-k multiple times.  In that case, the namespace will be published under each of those keywords.

.TP
\fB\-m \fITYPE:VALUE\fR, \fB\-\-meta=\fITYPE:VALUE\fR
For the main file (or directory), set the metadata of the given TYPE to the given VALUE.  Note that this will not add the respective VALUE to the set of keywords under which the file can be found.

.TP
\fB\-p \fIPRIORITY\fR, \fB\-\-prio=\fIPRIORITY\fR
Set the priority of the namespace advertisement (default: 365).  If the local database is full, GNUnet will discard the content with the lowest ranking.  Note that ranks change over time depending on popularity.  The default should be high enough to preserve the locally inserted content in favor of content that migrates from other peers.

.TP
\fB\-q\fR, \fB\-\-quiet\fR
Do not print the list of pseudonyms (only perform create or delete operation).

.TP
\fB\-r \fILEVEL\fR, \fB\-\-replication=\fILEVEL\fR
Set the desired replication level.  If CONTENT_PUSHING is set to YES, GNUnet will push each block (for the file) LEVEL times to other peers before doing nomral "random" replication of all content.  This option can be used to push some content out into the network harder. Note that pushing content LEVEL times into the network does not guarantee that there will actually be LEVEL replicas.

.TP
\fB\-R IDENTIFIER\fR, \fB\-\-root=IDENTIFIER\fR
Specify the identifier for the root of the namespace.  Used in the namespace advertisement to tell users that find the namespace advertisement about an entry\-point into the namespace (use with \-C).  Advertisements are only created if "\-C" and "\-r" are specified.

.TP
\fB\-s ID:VALUE\fR, \fB\-\-set-rating=ID:VALUE\fR
Change the rating for the namespace identified by ID by VALUE.  For example, "\-s test:-3" decrements the rating of the pseudonym "test" by 3.  Note that ratings are purely local.  Each user has his own independent rating of namespaces.  The rating is merely a way for each user to keep track of his own experience with a given namespace.

.SH FILES
.TP
~/.gnunet/data/pseudonym/
Directory where information about non-local pseudonyms is stored
.TP
~/.gnunet/fs/updates/
Directory where information about possible updates to local pseudonyms is stored

.SH "REPORTING BUGS"
Report bugs by using Mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet\-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet\-identity\fP(1), \fBgnunet\-publish\fP(1), \fBgnunet\-search\fP(1)


ChangeLog
Mon Oct 31 18:46:36 CET 2011
	Releasing GNUnet 0.9.0pre4.  This release includes a working
	implementation of mesh (unoptimized), a working implementation
	of the VPN (could use more testing), a working implementation 
	of the WLAN transport (still very slow, unoptimized) and
	many, many bugfixes all over the place.  The major known
	issues are:

	* TESTING is known to sometimes crash on shutdown
	* DV is known to crash / not work properly
	* CHAT is experimental and was not reviewed yet

	More details to individual bugs that were fixed in this
	release are documented at:
	https://gnunet.org/bugs/changelog_page.php?version_id=182

Mon Sep 19 14:16:22 CEST 2011
	Releasing GNUnet 0.9.0pre3.  Instead of listing all that has
	changed, here is the short list of what is known NOT to work,
	in order of minor flaws to major issues, by module:

	* FS still has a few missing features, but largely works
	* TRANSPORT ATS is not used for actual bandwidth assignment,
	          important options to make it truly useful are still missing
	* TESTING does not support incremental HELLO changes, 
	          code has known shutdown issues (can crash)
	* VPN is experimental, but has been shown to work to some degree
	* MESH does not work (mock API emulates for VPN)
        * TRANSPORT-WLAN is experimental
        * TRANSPORT-HTTP is experimental, has known bugs
	* TRANSPORT-DV is experimental, has known bugs, protocol may still change
	* CHAT is experimental, has known bugs and was not reviewed yet

Thu Dec 23 23:33:05 CET 2010
	Releasing GNUnet 0.9.0pre2.

Thu Oct	 7 20:51:05 CEST 2010
	Removed dependency on GNU MP (GMP). The code now uses the libgcrypt MPI
	interface.

Sat Jul  3 20:47:45 CEST 2010
	Releasing GNUnet 0.9.0pre1.

Wed Jun 23 16:34:38 CEST 2010
	Added support for systemd-compatible passing of listen-sockets
	by ARM to services as well as systemd compatibility for gnunet-service-arm
	itself.  At least for non-MINGW systems this should work.

Sat Jun  5 18:08:39 CEST 2010
	Added support for UNIX domain sockets, code also defaults to
	them when available.

Sun May  2 13:49:10 CEST 2010
	Fixed problem with platform-dependence of format for IP addresses
	in HELLOs for TCP and UDP transport.
	Reduced address length field in HELLOs to 16 bit (was 32 bit).
	These changes break transport compatibility.

Fri Apr 16 18:19:05 CEST 2010
	Nearly complete rewrite and new overall architecture.  Many
	features are still missing, but basic system seems to be
	working again.  Maybe time to again track major changes
	in the ChangeLog.  Releasing 0.9.0pre0.


Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent.
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).

Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in
	util/network/dns.c; improved IPv6 support for
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.


src/fs/gnunet-pseudonym.c
/*
     This file is part of GNUnet.
     (C) 2001-2013 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 3, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/
/**
 * @file fs/gnunet-pseudonym.c
 * @brief manage GNUnet namespaces / pseudonyms
 * @author Christian Grothoff
 */
#include "platform.h"
#include "gnunet_fs_service.h"
#include "gnunet_identity_service.h"


/**
 * -A option
 */
static char *advertise_ns;

/**
 * -k option
 */
static struct GNUNET_FS_Uri *ksk_uri;

/**
 * -m option.
 */
static struct GNUNET_CONTAINER_MetaData *adv_metadata;

/**
 * Our block options (-p, -r, -a).
 */
static struct GNUNET_FS_BlockOptions bo = { {0LL}, 1, 365, 1 };

/**
 * -q option given.
 */
static int no_remote_printing;

/**
 * -r option.
 */
static char *root_identifier;

/**
 * -s option.
 */
static char *rating_change;

/**
 * Handle to fs service.
 */
static struct GNUNET_FS_Handle *h;

/**
 * Our configuration.
 */
static const struct GNUNET_CONFIGURATION_Handle *cfg;

/**
 * Handle to identity service.
 */
static struct GNUNET_IDENTITY_Handle *identity;

/**
 * Target namespace.
 */
static struct GNUNET_IDENTITY_Ego *namespace;

/**
 * URI to advertise.
 */
static struct GNUNET_FS_Uri *sks_uri;

/**
 * Global return value.
 */ 
static int ret;


/**
 * Progress callback given to FS.
 * 
 * @param cls unused
 * @param info progress information, unused
 */ 
static void *
progress_cb (void *cls, const struct GNUNET_FS_ProgressInfo *info)
{
  return NULL;
}


/**
 * Output information about a pseudonym.
 *
 * @param cls closure
 * @param pseudonym hash code of public key of pseudonym
 * @param name name of the pseudonym (might be NULL)
 * @param unique_name unique name of the pseudonym (might be NULL)
 * @param md meta data known about the pseudonym
 * @param rating the local rating of the pseudonym
 * @return GNUNET_OK to continue iteration, GNUNET_SYSERR to abort
 */
static int
pseudo_printer (void *cls, 
		const struct GNUNET_CRYPTO_EccPublicKey *pseudonym,
                const char *name, 
		const char *unique_name,
                const struct GNUNET_CONTAINER_MetaData *md, 
		int32_t rating)
{
  char *id;
  char *unique_id;
  int getinfo_result;

  /* While we get a name from the caller, it might be NULL.
   * GNUNET_FS_pseudonym_get_info () never returns NULL.
   */
  getinfo_result = GNUNET_FS_pseudonym_get_info (cfg, pseudonym,
						 NULL, NULL, &id, NULL);
  if (GNUNET_OK != getinfo_result)
  {
    GNUNET_break (0);
    return GNUNET_OK;
  }
  unique_id = GNUNET_FS_pseudonym_name_uniquify (cfg, pseudonym, id, NULL);
  GNUNET_free (id);
  FPRINTF (stdout, 
	   "%s (%d):\n", 
	   unique_id, rating);
  GNUNET_CONTAINER_meta_data_iterate (md, &EXTRACTOR_meta_data_print, stdout);
  FPRINTF (stdout, 
	   "%s",
	   "\n");
  GNUNET_free (unique_id);
  return GNUNET_OK;
}


/**
 * Function called once advertising is finished.
 * 
 * @param cls closure (NULL)
 * @param uri the advertised URI
 * @param emsg error message, NULL on success
 */
static void
post_advertising (void *cls,
		  const struct GNUNET_FS_Uri *uri, 
		  const char *emsg)
{
  if (emsg != NULL)
  {
    FPRINTF (stderr, "%s", emsg);
    ret = 1;
  }
  GNUNET_FS_stop (h);
  GNUNET_IDENTITY_disconnect (identity);
}


/**
 * Function called by identity service with known pseudonyms.
 *
 * @param cls closure, NULL
 * @param ego ego handle
 * @param ctx context for application to store data for this ego
 *                 (during the lifetime of this process, initially NULL)
 * @param name name assigned by the user for this ego,
 *                   NULL if the user just deleted the ego and it
 *                   must thus no longer be used
 */
static void
identity_cb (void *cls, 
	     struct GNUNET_IDENTITY_Ego *ego,
	     void **ctx,
	     const char *name)
{
  char *emsg;
  struct GNUNET_CRYPTO_EccPublicKey pub;

  if (NULL == ego) 
  {
    if (NULL == namespace)
    {
      ret = 1;
      return;
    }
    if (NULL != root_identifier)
    {
      if (NULL == ksk_uri)
      {
	emsg = NULL;
	ksk_uri = GNUNET_FS_uri_parse ("gnunet://fs/ksk/namespace", &emsg);
	GNUNET_assert (NULL == emsg);
      }
      GNUNET_IDENTITY_ego_get_public_key (namespace,
					  &pub);
      sks_uri = GNUNET_FS_uri_sks_create (&pub,
					  root_identifier);
      GNUNET_FS_publish_ksk (h, ksk_uri, adv_metadata, sks_uri,
			     &bo,
			     GNUNET_FS_PUBLISH_OPTION_NONE,
			     &post_advertising, NULL);
      GNUNET_FS_uri_destroy (sks_uri);
      return;
    }
    else
    {
      if (NULL != ksk_uri)
	FPRINTF (stderr, _("Option `%s' ignored\n"), "-k");
      if (NULL != advertise_ns)
	FPRINTF (stderr, _("Option `%s' ignored\n"), "-A");
    }
    return;
  }
  if (0 == strcmp (name, advertise_ns))
    namespace = ego;
}


/**
 * Main function that will be run by the scheduler.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param c configuration
 */
static void
run (void *cls, char *const *args, const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *c)
{
  struct GNUNET_CRYPTO_EccPublicKey nsid;
  char *set;
  int delta;

  cfg = c;
  h = GNUNET_FS_start (cfg, "gnunet-pseudonym", &progress_cb, NULL,
                       GNUNET_FS_FLAGS_NONE, GNUNET_FS_OPTIONS_END);
  if (NULL != rating_change)
  {
    set = rating_change;
    while ((*set != '\0') && (*set != ':'))
      set++;
    if (*set != ':')
    {
      GNUNET_log (GNUNET_ERROR_TYPE_ERROR, 
		  _("Invalid argument `%s'\n"),
                  rating_change);
    }
    else
    {
      *set = '\0';
      delta = strtol (&set[1], NULL,    /* no error handling yet */
                      10);
      if (GNUNET_OK == GNUNET_FS_pseudonym_name_to_id (cfg, rating_change, &nsid))
      {
        (void) GNUNET_FS_pseudonym_rank (cfg, &nsid, delta);
      }
      else
      {
        GNUNET_log (GNUNET_ERROR_TYPE_ERROR, 
                    ("Namespace `%s' unknown. Make sure you specify its numeric suffix, if any.\n"),
                    rating_change);
      }
    }
    GNUNET_free (rating_change);
    rating_change = NULL;
  }
  if (0 == no_remote_printing)
    GNUNET_FS_pseudonym_list_all (cfg, &pseudo_printer, NULL);

  if (NULL != advertise_ns)
    identity = GNUNET_IDENTITY_connect (cfg, 
					&identity_cb, 
					NULL);
}


/**
 * The main function to manipulate GNUnet pseudonyms (and publish
 * to namespaces).
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc, char *const *argv)
{
  static const struct GNUNET_GETOPT_CommandLineOption options[] = {
    {'a', "anonymity", "LEVEL",
     gettext_noop ("set the desired LEVEL of sender-anonymity"),
     1, &GNUNET_GETOPT_set_uint, &bo.anonymity_level},
    {'A', "advertise", "NAME",
     gettext_noop ("advertise namespace NAME"),
     1, &GNUNET_GETOPT_set_string, &advertise_ns},
    {'k', "keyword", "VALUE",
     gettext_noop ("add an additional keyword for the advertisment"
                   " (this option can be specified multiple times)"),
     1, &GNUNET_FS_getopt_set_keywords, &ksk_uri},
    {'m', "meta", "TYPE:VALUE",
     gettext_noop ("set the meta-data for the given TYPE to the given VALUE"),
     1, &GNUNET_FS_getopt_set_metadata, &adv_metadata},
    {'p', "priority", "PRIORITY",
     gettext_noop ("use the given PRIORITY for the advertisments"),
     1, &GNUNET_GETOPT_set_uint, &bo.content_priority},
    {'q', "quiet", NULL,
     gettext_noop ("do not print names of remote namespaces"),
     0, &GNUNET_GETOPT_set_one, &no_remote_printing},
    {'r', "replication", "LEVEL",
     gettext_noop ("set the desired replication LEVEL"),
     1, &GNUNET_GETOPT_set_uint, &bo.replication_level},
    {'R', "root", "ID",
     gettext_noop ("specify ID of the root of the namespace"),
     1, &GNUNET_GETOPT_set_string, &root_identifier},
    {'s', "set-rating", "ID:VALUE",
     gettext_noop ("change rating of namespace ID by VALUE"),
     1, &GNUNET_GETOPT_set_string, &rating_change},
    GNUNET_GETOPT_OPTION_END
  };
  bo.expiration_time =
      GNUNET_FS_year_to_time (GNUNET_FS_get_current_year () + 2);

  if (GNUNET_OK != GNUNET_STRINGS_get_utf8_args (argc, argv, &argc, &argv))
    return 2;

  ret = (GNUNET_OK ==
	 GNUNET_PROGRAM_run (argc, argv, "gnunet-pseudonym [OPTIONS]",
			     gettext_noop ("Manage GNUnet pseudonyms."),
			     options, &run, NULL)) ? ret : 1;
  GNUNET_free ((void*) argv);
  return ret;
}

/* end of gnunet-pseudonym.c */
